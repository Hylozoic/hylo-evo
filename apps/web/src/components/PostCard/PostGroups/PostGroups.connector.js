import { connect } from 'react-redux'

export function mapStateToProps (state, props) {
  let isPublic = props.isPublic
  let groupsPlusPublic = props.groups

  if (isPublic) {
    groupsPlusPublic.unshift({ name: 'Public', id: 'public', avatarUrl: '/public-icon.svg', slug: 'public' })
  }

  return {
    groups: groupsPlusPublic
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {}
}

export default connect(mapStateToProps, mapDispatchToProps)
