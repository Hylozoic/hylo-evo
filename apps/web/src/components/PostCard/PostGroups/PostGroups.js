import React, { Component } from 'react'
import { useTranslation, withTranslation } from 'react-i18next'
import { get, isEmpty } from 'lodash/fp'
import cx from 'classnames'
import { Link } from 'react-router-dom'
import { groupUrl } from 'util/navigation'
import GroupsList from 'components/GroupsList'
import Icon from 'components/Icon'
import classes from './PostGroups.module.scss'

class PostGroups extends Component {
  static defaultState = {
    expanded: false
  }

  constructor (props) {
    super(props)
    this.state = PostGroups.defaultState
  }

  toggleExpanded = () => {
    this.setState({
      expanded: !this.state.expanded
    })
  }

  render () {
    const { groups, constrained, slug, showBottomBorder, t } = this.props
    const { expanded } = this.state

    // don't show if there are no groups or this isn't cross posted
    if (isEmpty(groups) || (groups.length === 1 && get('0.slug', groups) === slug)) return null

    return <div className={cx(classes.groups, { [classes.constrained]: constrained, [classes.expanded]: expanded, [classes.bottomBorder]: showBottomBorder })} onClick={expanded ? this.toggleExpanded : undefined}>
      <div className={classes.row}>
        <span className={classes.label}>{`${this.props.t('Posted In:')} `}</span>
        {!expanded &&
          <LinkedGroupNameList t={t} groups={groups} maxShown={2} expandFunc={this.toggleExpanded} />}
        <a onClick={this.toggleExpanded} className={classes.expandLink}><Icon name={expanded ? 'ArrowUp' : 'ArrowDown'} className={classes.expandIcon} /></a>
      </div>

      {expanded && <GroupsList groups={groups} />}
    </div>
  }
}

export function LinkedGroupNameList ({ groups, maxShown = 2, expandFunc, t }) {
  const groupsToDisplay = (maxShown && maxShown <= groups.length)
    ? groups.slice(0, maxShown)
    : groups
  const othersCount = groups.length - groupsToDisplay.length

  return <span className={classes.groupList}>
    {groupsToDisplay.map((group, i) =>
      <LinkedGroupName group={group} key={i}>
        <Separator currentIndex={i} displayCount={groupsToDisplay.length} t={t} othersCount={othersCount} />
      </LinkedGroupName>)}
    {othersCount > 0 &&
      <Others othersCount={othersCount} expandFunc={expandFunc} />}
  </span>
}

export function LinkedGroupName ({ group, children }) {
  return <span key={group.id}>
    <Link to={groupUrl(group.slug)} className={classes.groupLink}>{group.name === 'Public' && <Icon name='Public' className={classes.publicGroupIcon} />} {group.name}</Link>
    {children}
  </span>
}

export function Separator ({ currentIndex, displayCount, othersCount, t }) {
  const isLastEntry = currentIndex === displayCount - 1
  const isNextToLastEntry = currentIndex === Math.max(0, displayCount - 2)
  const hasOthers = othersCount > 0

  if (isLastEntry) return null
  if (!hasOthers && isNextToLastEntry) return <span key='and'> {t('and')} </span>

  return <span>, </span>
}

export function Others ({ othersCount, expandFunc }) {
  const { t } = useTranslation()
  if (othersCount < 0) return null

  const phrase = othersCount === 1 ? t('1 other') : t('{{othersCount}} others', { othersCount })

  return <React.Fragment>
    <span key='and'> {t('and')} </span>
    <a key='others' className={classes.groupLink} onClick={expandFunc}>{phrase}</a>
  </React.Fragment>
}

export default withTranslation()(PostGroups)
