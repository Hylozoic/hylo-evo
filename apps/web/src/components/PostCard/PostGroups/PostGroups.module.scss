.groups {
  padding-right: 0;
  display: flex;
  justify-content: flex-start;
  border-bottom: 1px solid rgba(235, 235, 235, 1.0);
  overflow: hidden;
  min-height: 40px;
  background-color: rgba(246, 247, 249, 1.0);
}

.constrained.groups {
}

.groupList {
  composes: timestamp from 'css/typography.scss';
  overflow: hidden;
  text-overflow: ellipsis;
}

.groupLink {
  color: $color-caribbean-green;
  font-size: 12px;
  &:hover {
    text-decoration: none;
    color: $color-gossamer;
  }
}

.bottomBorder {
}

.expanded {
  padding-top: 7px;
  width: 100%;
  padding-bottom: 10px;
  flex-wrap: wrap;
}

.row {
  display: flex;
  align-items: center;
  width: 100%;
  padding: 0 8px;
}

.label {
  composes: timestamp from 'css/typography.scss';
}

.publicGroupIcon {
  color: $color-caribbean-green;
  padding-left: 3px;
  font-size: 14px;
  top: 2px;
}

.expandLink {
  margin-left: 5px;
  margin-right: 22px;
  display: flex;
  align-items: center;
}

.expandIcon {
  composes: timestamp from 'css/typography.scss';
  font-size: 20px;
}

.expandedSection {
  width: 100%;
  padding-bottom: 10px;
}

.groupRow {
  display: flex;
  margin-bottom: 5px;
}

.groupCell {
  flex: 1;
  &:hover {
    text-decoration: none;
  }
}

.groupCellAvatar {
  @include image(20px);
  display: inline-block;
  vertical-align: middle;
  border-radius: 4px;
  margin-right: 10px;
}

.groupCellName {
  composes: text-button from 'css/typography.scss';
  font-size: 12px;
}

@media screen and (max-width: 325px) {

  .groupList {
    overflow: hidden;
    display: block;
    height: 20px;
    text-overflow: ellipsis;
    line-height: 20px;
    padding-left: 5px;
  }

  .label {
    text-align: right;
  }

}
