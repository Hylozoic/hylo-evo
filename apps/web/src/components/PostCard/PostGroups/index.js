import component from './PostGroups'
import connector from './PostGroups.connector'

export default connector(component)
