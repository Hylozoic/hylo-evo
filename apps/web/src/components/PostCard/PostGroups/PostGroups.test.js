import React from 'react'
import PostGroups from './PostGroups'
import { shallow } from 'enzyme'

describe('PostGroups', () => {
  it('matches last snapshot', () => {
    const props = {
      groups: [
        {
          id: 1,
          name: 'One',
          slug: 'one'
        },
        {
          id: 2,
          name: 'Two',
          slug: 'two'
        },
        {
          id: 3,
          name: 'Three',
          slug: 'three'
        }
      ],
      slug: 'hylo'
    }

    const wrapper = shallow(<PostGroups {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('matches last snapshot when expanded', () => {
    const props = {
      groups: [
        {
          id: 1,
          name: 'One',
          slug: 'one'
        },
        {
          id: 2,
          name: 'Two',
          slug: 'two'
        },
        {
          id: 3,
          name: 'Three',
          slug: 'three'
        }
      ],
      slug: 'hylo'
    }
    const wrapper = shallow(<PostGroups {...props} />)
    wrapper.instance().toggleExpanded()
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })

  it('returns null when in the only group', () => {
    const props = {
      groups: [
        {
          id: 1,
          name: 'One',
          slug: 'one'
        }
      ],
      slug: 'one'
    }

    const wrapper = shallow(<PostGroups {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
