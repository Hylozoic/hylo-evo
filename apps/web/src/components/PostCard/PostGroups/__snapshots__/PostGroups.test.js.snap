// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostGroups matches last snapshot 1`] = `
<div
  data-stylename="groups"
>
  <div
    data-stylename="row"
  >
    <span
      data-stylename="label"
    >
      Posted In: 
    </span>
    <LinkedGroupNameList
      expandFunc={[Function]}
      groups={
        Array [
          Object {
            "id": 1,
            "name": "One",
            "slug": "one",
          },
          Object {
            "id": 2,
            "name": "Two",
            "slug": "two",
          },
          Object {
            "id": 3,
            "name": "Three",
            "slug": "three",
          },
        ]
      }
      maxShown={2}
      t={[Function]}
    />
    <a
      data-stylename="expandLink"
      onClick={[Function]}
    >
      <Icon
        data-stylename="expandIcon"
        name="ArrowDown"
      />
    </a>
  </div>
</div>
`;

exports[`PostGroups matches last snapshot when expanded 1`] = `
<div
  data-stylename="groups expanded"
  onClick={[Function]}
>
  <div
    data-stylename="row"
  >
    <span
      data-stylename="label"
    >
      Posted In: 
    </span>
    <a
      data-stylename="expandLink"
      onClick={[Function]}
    >
      <Icon
        data-stylename="expandIcon"
        name="ArrowUp"
      />
    </a>
  </div>
  <GroupsList
    groups={
      Array [
        Object {
          "id": 1,
          "name": "One",
          "slug": "one",
        },
        Object {
          "id": 2,
          "name": "Two",
          "slug": "two",
        },
        Object {
          "id": 3,
          "name": "Three",
          "slug": "three",
        },
      ]
    }
  />
</div>
`;

exports[`PostGroups returns null when in the only group 1`] = `""`;
